zookeeper-server/src/main/java/org/apache/zookeeper/server/util/QuotaMetricsUtils.java

```diff
@@ -0,0 +1,158 @@
+        final Map<String, Number> metricsMap = new ConcurrentHashMap<>();
+        if (dataTree != null) {
+            getQuotaLimitOrUsage(Quotas.quotaZookeeper, metricsMap, type, dataTree);
+        }
+        return metricsMap;
+    }
+
+    static void getQuotaLimitOrUsage(final String path,
+                                     final Map<String, Number> metricsMap,
+                                     final QUOTA_LIMIT_USAGE_METRIC_TYPE type,
+                                     final DataTree dataTree) {
+        final DataNode node = dataTree.getNode(path);
+        if (node == null) {
+            return;
+        }
+        final Set<String> children = node.getChildren();
+        if (children.isEmpty()) {
+            if (shouldCollect(path, type)) {
+                collectQuotaLimitOrUsage(path, node, metricsMap, type);
+            }
+            return;
+        }
+        for (final String child : children) {
+            getQuotaLimitOrUsage(path + "/" + child, metricsMap, type, dataTree);
+        }
+    }
+
+    static boolean shouldCollect(final String path, final QUOTA_LIMIT_USAGE_METRIC_TYPE type) {
+        return path.endsWith(LIMIT_END_STRING)
+                && (QUOTA_LIMIT_USAGE_METRIC_TYPE.QUOTA_COUNT_LIMIT == type || QUOTA_LIMIT_USAGE_METRIC_TYPE.QUOTA_BYTES_LIMIT == type)
+                || path.endsWith(STATS_END_STRING)
+                && (QUOTA_LIMIT_USAGE_METRIC_TYPE.QUOTA_COUNT_USAGE == type || QUOTA_LIMIT_USAGE_METRIC_TYPE.QUOTA_BYTES_USAGE == type);
+    }
+
+    static void collectQuotaLimitOrUsage(final String path,
+                                         final DataNode node,
+                                         final Map<String, Number> metricsMap,
+                                         final QUOTA_LIMIT_USAGE_METRIC_TYPE type) {
+        final String namespace = PathUtils.getTopNamespace(Quotas.trimQuotaPath(path));
+        if (namespace == null) {
+            return;
+        }
+        if (node.getData() == null) {
+            return;
+        }
+        final StatsTrack statsTrack = new StatsTrack(node.getData());
+        switch (type) {
+            case QUOTA_COUNT_LIMIT:
+                metricsMap.put(namespace, getQuotaLimit(statsTrack.getCountHardLimit(), statsTrack.getCount()));
+                break;
+            case QUOTA_BYTES_LIMIT:
+                metricsMap.put(namespace, getQuotaLimit(statsTrack.getByteHardLimit(), statsTrack.getBytes()));
+                break;
+            case QUOTA_COUNT_USAGE:
+                metricsMap.put(namespace, statsTrack.getCount());
+                break;
+            case QUOTA_BYTES_USAGE:
+                metricsMap.put(namespace, statsTrack.getBytes());
+                break;
+            default:
+        }
+    }
+
+    // hard limit takes precedence if specified
+    static long getQuotaLimit(final long hardLimit, final long limit) {
+        return hardLimit > -1 ? hardLimit : limit;
+    }
```

Comment:
This is another case of policy and mechanism getting mixed up; do we really want to prevent people from monitoring their soft quota errors?

Reply:
Hmm, not sure what you meant. This is the limit used for both soft quota errors and hard quota errors.  If hardLimit is specified, then the client will get QuotaExceededException, otherwise it just log as WARN.

Reply:
Sorry---I was indeed too quick (and not very clear to boot).

> We are not preventing people from monitoring the soft quota errors. In fact, we publish the \"proper\" metric data so they can set up the alerts based on different usage level.
>
> The limit is published in this way because
>
> -   It matches up with what\'s been used in the checkQuota() logic, so user doesn\'t need to know how the soft limit and hard limit are used internally.
> -   This is the limit used by the server to check quota and this is what they need to monitor and alert on. It can be soft limit or hard limit.
> -   In the future if the quota check logic is changed in term of how the soft limit and hard limit are used, we just need to change the server side code. User\'s monitoring and alert rule do not need to be changed.

You are right.

I had in mind that soft/hard quotas had the same semantics as, e.g., disk quotas under Unix---but that is not what is implemented as of today.

(That being said: it is something I would like to see improved at some point—I believe the current semantics are accidental, and perhaps not as useful as they could be—but my wishes do no make reality.

IMO, server administrators should be informed when a node is above the soft quota, even if no errors are generated. Similarly, clients should have a way of monitoring the \"warnings\" which are being issued against them.

So it seems that published metrics will have to be updated when specifying both a soft and hard limit on a node becomes useful—e.g., by adding `quota_count_soft_limit_per_namespace` and `quota_bytes_soft_limit_per_namespace`—so that dashboards can show watermarks as actually configured in the quota subsystem.)

Reply:
Yes, if we change the semantics of soft and hard quotas in the future, we will update the the published metrics along with the changes.  Changing the semantics is a fundamental thing, so it makes sense to make all the changes that are impacted at that time.

There are two concerns for publishing both soft and hard limit at this point.

1. It doesn't match with the current semantics of soft and hard quotas and make it harder for user t o use the metrics. 
2. Given the constraint we have with the current interface of metrics framework, it means the we need to iterate two more times on the /zookeeper/quota tree. IMO, we pay something that doesn't give us too much benefits.

Reply:
I agree.